File: go/include.go
package entryconf

import (
	"errors"
	"path/filepath"
	"strconv"
	"strings"
)

const (
	includePrefix = "@file:"

	// maxIncludeDepth bounds how deeply includes may nest.
	maxIncludeDepth = 100
)

// walk is the position of the include resolver: the file whose value is being
// walked and its directory (paths are relative to the referencing file), the
// pointer within that file (src) and within the effective tree (eff), the
// stack of files currently being included for cycle detection (innermost
// last), and the @file: references traversed so far (SPEC §10.3 "chain").
type walk struct {
	doc   string
	dir   string
	src   string
	eff   string
	files []string
	chain []Reference
}

func (w *walk) step(token string) *walk {
	next := *w
	next.src = w.src + "/" + escapePointerToken(token)
	next.eff = w.eff + "/" + escapePointerToken(token)
	return &next
}

// resolveIncludes walks a parsed document and replaces every "@file:<path>"
// string with the parsed tree of the referenced file (SPEC §5).
func (l *loader) resolveIncludes(v any, w *walk) (any, error) {
	switch t := v.(type) {
	case string:
		return l.resolveIncludeString(t, w)
	case map[string]any:
		out := make(map[string]any, len(t))
		for k, val := range t {
			// Keys are never includes and are never interpolated (SPEC §6).
			resolved, err := l.resolveIncludes(val, w.step(k))
			if err != nil {
				return nil, err
			}
			out[k] = resolved
		}
		return out, nil
	case []any:
		out := make([]any, len(t))
		for i, val := range t {
			resolved, err := l.resolveIncludes(val, w.step(strconv.Itoa(i)))
			if err != nil {
				return nil, err
			}
			out[i] = resolved
		}
		return out, nil
	default:
		return v, nil
	}
}

func (l *loader) resolveIncludeString(s string, w *walk) (any, error) {
	dir, chain := w.dir, w.files
	if !strings.HasPrefix(s, "@") {
		return s, nil
	}
	// SPEC §5 escaping: a leading "@@" becomes a literal "@" and the string is
	// never treated as an include.
	if strings.HasPrefix(s, "@@") {
		return "@" + s[2:], nil
	}
	if !strings.HasPrefix(s, includePrefix) {
		// Reserved for future directives.
		return nil, errf(CodeSubstitution, "unknown directive %q (write %q to mean a literal leading @)", s, "@"+s)
	}

	abs := includeTarget(s, dir)

	if _, ok := parserFor(abs); !ok {
		return nil, errf(CodeInclude, "unsupported include extension: %q", s)
	}
	for _, seen := range chain {
		if seen == abs {
			return nil, errf(CodeIncludeCycle, "include cycle: %s", strings.Join(append(append([]string{}, chain...), abs), " -> "))
		}
	}
	// Backstop for a cycle that path comparison cannot see, e.g. one made of
	// symlinks pointing at each other under different names.
	if len(chain) > maxIncludeDepth {
		return nil, errf(CodeIncludeCycle, "include nesting deeper than %d files: %s",
			maxIncludeDepth, strings.Join(append(append([]string{}, chain...), abs), " -> "))
	}

	doc, err := l.parseDocument(abs)
	if err != nil {
		var ecErr *Error
		if errors.As(err, &ecErr) {
			return nil, ecErr // E_PARSE for an unparseable target
		}
		return nil, wrapf(CodeInclude, err, "cannot read include target %q", abs)
	}
	next := make([]string, len(chain), len(chain)+1)
	copy(next, chain)
	refs := make([]Reference, len(w.chain), len(w.chain)+1)
	copy(refs, w.chain)
	refs = append(refs, Reference{Document: w.doc, Pointer: w.src})
	if l.rec != nil {
		l.rec.graft(abs, Graft{Effective: w.eff, Chain: refs})
	}
	return l.resolveIncludes(doc, &walk{doc: abs, dir: filepath.Dir(abs), src: "", eff: w.eff, files: append(next, abs), chain: refs})
}

// includeTarget resolves the path of an "@file:<path>" string relative to the
// directory of the file holding it (SPEC §5), cleaned.
func includeTarget(s, dir string) string {
	target := s[len(includePrefix):]
	abs := target
	if !filepath.IsAbs(abs) {
		abs = filepath.Join(dir, target)
	}
	return filepath.Clean(abs)
}

// isInclude reports whether an authored string is an "@file:" reference (as
// opposed to an "@@" escape, a literal, or a reserved directive).
func isInclude(s string) bool {
	return strings.HasPrefix(s, includePrefix)
}

File: rust/src/include.rs
//! SPEC §5 — `@file:` include grafting and `@@` escaping.

use std::fs;
use std::path::{Path, PathBuf};

use serde_json::{Map, Value};

use crate::error::{Error, ErrorCode};
use crate::parse;
use crate::pointer::escape_token;
use crate::source::{include_target, RawGraft};
use crate::Loader;

pub(crate) const INCLUDE_PREFIX: &str = "@file:";

/// Whether an authored string is an `@file:` reference (not an `@@` escape, a
/// literal, or a reserved directive).
pub(crate) fn is_include(s: &str) -> bool {
    s.starts_with(INCLUDE_PREFIX)
}

/// The resolver's position: the file whose value is being walked (lexical
/// path) and its directory, the pointer within that file and within the
/// effective tree, the canonical paths of the files being resolved (for cycle
/// detection, innermost last), and the `@file:` references traversed so far
/// (SPEC §10.3's chain, outermost first).
pub(crate) struct Walk {
    pub(crate) doc: PathBuf,
    pub(crate) dir: PathBuf,
    pub(crate) src: String,
    pub(crate) eff: String,
    pub(crate) stack: Vec<PathBuf>,
    pub(crate) chain: Vec<(PathBuf, String)>,
}

impl Walk {
    fn step(&self, token: &str) -> Walk {
        let suffix = format!("/{}", escape_token(token));
        Walk {
            doc: self.doc.clone(),
            dir: self.dir.clone(),
            src: format!("{}{suffix}", self.src),
            eff: format!("{}{suffix}", self.eff),
            stack: self.stack.clone(),
            chain: self.chain.clone(),
        }
    }
}

/// Resolves every `@file:` reference in `value` (SPEC §5).
pub(crate) fn resolve(l: &Loader<'_>, value: Value, w: &Walk) -> Result<Value, Error> {
    match value {
        Value::String(s) => resolve_string(l, s, w),
        Value::Array(items) => {
            let mut out = Vec::with_capacity(items.len());
            for (i, item) in items.into_iter().enumerate() {
                out.push(resolve(l, item, &w.step(&i.to_string()))?);
            }
            Ok(Value::Array(out))
        }
        Value::Object(object) => {
            let mut out = Map::new();
            for (key, item) in object {
                let resolved = resolve(l, item, &w.step(&key))?;
                out.insert(key, resolved);
            }
            Ok(Value::Object(out))
        }
        scalar => Ok(scalar),
    }
}

fn resolve_string(l: &Loader<'_>, s: String, w: &Walk) -> Result<Value, Error> {
    // `@@` is the escape and must be tested before the include prefix.
    if let Some(rest) = s.strip_prefix("@@") {
        return Ok(Value::String(format!("@{rest}")));
    }
    if let Some(target) = s.strip_prefix(INCLUDE_PREFIX) {
        return graft(l, target, w);
    }
    if s.starts_with('@') {
        return Err(Error::new(
            ErrorCode::Substitution,
            format!("{s:?} starts with `@` but is not an include; the `@` namespace is reserved (write `@@` for a literal `@`)"),
        ));
    }
    Ok(Value::String(s))
}

fn graft(l: &Loader<'_>, target: &str, w: &Walk) -> Result<Value, Error> {
    if target.is_empty() {
        return Err(Error::new(
            ErrorCode::Include,
            "`@file:` with an empty path".to_string(),
        ));
    }
    let joined = include_target(target, &w.dir);

    let Some(format) = parse::format_for_path(&joined) else {
        return Err(Error::new(
            ErrorCode::Include,
            format!(
                "include {target:?}: unsupported extension (want .json, .yaml, .yml, or .toml)"
            ),
        ));
    };

    // Cycle detection compares canonical paths, so a file reached under two
    // names (a symlink, a `..` detour) is still one file.
    let canonical = fs::canonicalize(&joined).map_err(|e| {
        Error::new(
            ErrorCode::Include,
            format!("include {target:?} ({}): {e}", joined.display()),
        )
    })?;
    if !l.src.is_file(&joined) {
        return Err(Error::new(
            ErrorCode::Include,
            format!(
                "include {target:?} ({}) is not a regular file",
                joined.display()
            ),
        ));
    }

    if let Some(at) = w.stack.iter().position(|p| *p == canonical) {
        let mut chain: Vec<String> = w.stack[at..]
            .iter()
            .map(|p| p.display().to_string())
            .collect();
        chain.push(canonical.display().to_string());
        return Err(Error::new(
            ErrorCode::IncludeCycle,
            format!("include cycle: {}", chain.join(" -> ")),
        ));
    }

    let bytes = l.src.read_file(&joined).map_err(|e| {
        Error::new(
            ErrorCode::Include,
            format!("include {target:?} ({}): {e}", joined.display()),
        )
    })?;
    let text = String::from_utf8(bytes.clone()).map_err(|_| {
        Error::new(
            ErrorCode::Parse,
            format!("{} is not valid UTF-8", joined.display()),
        )
    })?;
    let tree = parse::parse(&text, format, &joined)?;
    let mut refs = w.chain.clone();
    refs.push((w.doc.clone(), w.src.clone()));
    if let Some(rec) = l.rec {
        rec.record(&joined, &bytes, format.into(), Some(tree.clone()));
        rec.graft(
            &joined,
            RawGraft {
                effective: w.eff.clone(),
                chain: refs.clone(),
            },
        );
    }

    let parent = joined
        .parent()
        .map_or_else(|| PathBuf::from("."), Path::to_path_buf);
    let mut stack = w.stack.clone();
    stack.push(canonical);
    let next = Walk {
        doc: joined,
        dir: parent,
        src: String::new(),
        eff: w.eff.clone(),
        stack,
        chain: refs,
    };
    resolve(l, tree, &next)
}
